flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -285,7 +285,12 @@ private ExecutionGraphInfo createExecutionGraphInfoWithJobStatus(JobStatus jobSt
     }
 
     private void jobAlreadyDone() {
-        resultFuture.completeExceptionally(new JobNotFinishedException(getJobID()));
+        resultFuture.complete(
+                JobManagerRunnerResult.forSuccess(
+                        new ExecutionGraphInfo(
+                                jobMasterServiceProcessFactory.createArchivedExecutionGraph(
+                                        JobStatus.FAILED,
+                                        new JobAlreadyDoneException(getJobID())))));
```

Comment:
I think if we communicate that way to the `Dispatcher` that a job has been already terminated, then we don't need `JobManagerRunnerResult.isJobAlreadyDone()`.